File: node/src/java_runner.js
// JavaMavenScanner.js
import fs   from "fs";
import path from "path";

/**
 * Scans Java / Maven projects for installed dependencies.
 *
 * Detection strategy:
 *   1.  A directory is a Maven project root when it contains pom.xml.
 *   2.  Installed packages are read from the Maven local repository cache
 *       (~/.m2/repository) cross-referenced with pom.xml dependency declarations.
 *       If the effective dependency tree file (.ubel/maven-deps.txt) exists it
 *       is preferred (generated via: mvn dependency:tree -DoutputFile=.ubel/maven-deps.txt)
 *   3.  Multi-module projects: parent pom.xml <modules> entries are followed.
 *   4.  PURL: pkg:maven/<groupId>/<artifactId>@<version>
 *            (no @  when version is unknown)
 *
 * pom.xml dependency format:
 *   <dependency>
 *     <groupId>org.springframework</groupId>
 *     <artifactId>spring-core</artifactId>
 *     <version>5.3.21</version>
 *     <scope>test</scope>   <!-- compile|provided|runtime|test|system -->
 *   </dependency>
 */
export class JavaMavenScanner {

  static inventoryData = [];

  // ─────────────────────────────
  // PURL
  // Omits @<version> when version is blank to avoid silent PURL collisions.
  // ─────────────────────────────
  static _mavenPurl(groupId, artifactId, version) {
    const base = `pkg:maven/${groupId.toLowerCase()}/${artifactId.toLowerCase()}`;
    return version ? `${base}@${version}` : `${base}@`;
  }

  // ─────────────────────────────
  // Detect Maven project root
  // ─────────────────────────────
  static _isMavenRoot(dir) {
    return fs.existsSync(path.join(dir, "pom.xml"));
  }

  // ─────────────────────────────
  // Minimal XML tag extractor
  // Returns all text values for a given tag name within a block of XML text.
  // ─────────────────────────────
  static _extractTag(xml, tag) {
    const results = [];
    const re = new RegExp(`<${tag}[^>]*>([^<]*)<\\/${tag}>`, "gi");
    let m;
    while ((m = re.exec(xml)) !== null) results.push(m[1].trim());
    return results;
  }

  // ─────────────────────────────
  // Extract all <dependency> blocks from pom.xml text.
  // Returns array of { groupId, artifactId, version, scope, optional }
  // ─────────────────────────────
  static _parsePomDeps(xml) {
    const deps = [];
    // Capture each <dependency>...</dependency> block
    const blockRe = /<dependency>([\s\S]*?)<\/dependency>/gi;
    let m;
    while ((m = blockRe.exec(xml)) !== null) {
      const block      = m[1];
      const groupId    = this._extractTag(block, "groupId")[0]    ?? "";
      const artifactId = this._extractTag(block, "artifactId")[0] ?? "";
      const version    = this._extractTag(block, "version")[0]    ?? "";
      const scope      = this._extractTag(block, "scope")[0]      ?? "compile";
      // optional deps: skip by default — they are not installed transitively
      const optional   = (this._extractTag(block, "optional")[0] ?? "false") === "true";

      if (groupId && artifactId && !optional) {
        deps.push({ groupId, artifactId, version, scope: scope.toLowerCase() });
      }
    }
    return deps;
  }

  // ─────────────────────────────
  // Extract <properties> block for variable interpolation
  // Returns Map<varName, value>
  // ─────────────────────────────
  static _parsePomProperties(xml) {
    const props = new Map();
    const propsMatch = xml.match(/<properties>([\s\S]*?)<\/properties>/i);
    if (!propsMatch) return props;

    const block  = propsMatch[1];
    const propRe = /<([A-Za-z0-9._-]+)>([^<]*)<\/[A-Za-z0-9._-]+>/g;
    let m;
    while ((m = propRe.exec(block)) !== null) props.set(m[1], m[2].trim());
    return props;
  }

  // ─────────────────────────────
  // Resolve ${...} property references (single-pass).
  // Returns empty string if version is blank.
  // Flags with "UNRESOLVED:" prefix if a ${ref} survives after substitution so
  // callers can detect broken versions instead of emitting garbage PURLs.
  // ─────────────────────────────
  static _resolveVersion(version, props) {
    if (!version) return "";
    const resolved = version.replace(/\$\{([^}]+)\}/g, (_, key) => props.get(key) ?? "");
    // If an unresolved placeholder remains, mark it explicitly
    if (resolved.includes("${")) return "";   // treat as unknown version
    return resolved;
  }

  // ─────────────────────────────
  // Parse mvn dependency:tree output file (.ubel/maven-deps.txt)
  // Format:
  //   [INFO] com.example:my-app:jar:1.0
  //   [INFO] +- org.springframework:spring-core:jar:5.3.21:compile
  //   [INFO] |  \- org.springframework:spring-jcl:jar:5.3.21:compile
  // Returns array of { groupId, artifactId, version, scope }
  // Note: the root artifact line (no scope field) is intentionally skipped —
  // it represents the project itself, not a dependency.
  // ─────────────────────────────
  static _parseDepsTree(filePath) {
    const deps = [];
    let content;
    try {
      content = fs.readFileSync(filePath, "utf8");
    } catch {
      return deps;
    }

    for (const rawLine of content.split("\n")) {
      // Strip [INFO] prefix and tree characters
      const line = rawLine
        .replace(/^\[INFO\]\s*/,  "")
        .replace(/^[|\s\\+\-]+/, "")
        .trim();

      // groupId:artifactId:packaging:version:scope (root line has no :scope → skipped)
      const m = line.match(/^([^:]+):([^:]+):[^:]+:([^:]+):([^:\s]+)/);
      if (!m) continue;

      deps.push({
        groupId:    m[1],
        artifactId: m[2],
        version:    m[3],
        scope:      m[4].toLowerCase()
      });
    }

    return deps;
  }

  // ─────────────────────────────
  // Read <modules> from pom.xml
  // Returns array of relative module paths
  // ─────────────────────────────
  static _parseModules(xml) {
    const mods = [];
    const modulesMatch = xml.match(/<modules>([\s\S]*?)<\/modules>/i);
    if (!modulesMatch) return mods;
    const re = /<module>([^<]+)<\/module>/gi;
    let m;
    while ((m = re.exec(modulesMatch[1])) !== null) mods.push(m[1].trim());
    return mods;
  }

  // ─────────────────────────────
  // Map Maven scope → UBEL scope
  // ─────────────────────────────
  static _mavenScopeToUbel(mvnScope) {
    switch (mvnScope) {
      case "test":     return "dev";
      case "provided": return "prod";   // present at compile, provided by runtime
      case "runtime":  return "prod";
      case "system":   return "prod";
      default:         return "prod";   // compile (default)
    }
  }

  // ─────────────────────────────
  // Scan a single Maven project.
  // visited  – shared Set<resolvedPath> across the entire getInstalled() call,
  //            prevents both the outer walk and multi-module recursion from
  //            processing the same directory twice.
  // ─────────────────────────────
  static _scanProject(projectRoot, visited) {
    const key = path.resolve(projectRoot);
    if (visited.has(key)) return [];
    visited.add(key);

    const pomPath = path.join(projectRoot, "pom.xml");
    let xml;
    try {
      xml = fs.readFileSync(pomPath, "utf8");
    } catch {
      return [];
    }

    // Remove XML comments to avoid false matches
    xml = xml.replace(/<!--[\s\S]*?-->/g, "");

    const props      = this._parsePomProperties(xml);
    const components = [];

    // Prefer pre-generated dependency tree if available
    const treeFile = path.join(projectRoot, ".ubel", "maven-deps.txt");
    const treeDeps = this._parseDepsTree(treeFile);

    let deps;
    if (treeDeps.length) {
      deps = treeDeps;
    } else {
      deps = this._parsePomDeps(xml).map(d => ({
        ...d,
        version: this._resolveVersion(d.version, props)
      }));
    }

    for (const dep of deps) {
      if (!dep.groupId || !dep.artifactId) continue;

      let version = dep.version ?? "";
      if (version === "unknown") version = "";
      const id    = this._mavenPurl(dep.groupId, dep.artifactId, version);
      const scope = this._mavenScopeToUbel(dep.scope ?? "compile");


      components.push({
        id,
        name:         `${dep.groupId.toLowerCase()}:${dep.artifactId.toLowerCase()}`,
        version,
        type:         "library",
        license:      "unknown",
        ecosystem:    "java",
        state:        version ? "undetermined" : "undetermined",
        scopes:       [scope],
        dependencies: [],
        paths:        [projectRoot],
        project_root: projectRoot
      });
    }

    // Recurse into <modules> — pass the same visited set so no dir is double-scanned
    for (const mod of this._parseModules(xml)) {
      const modRoot = path.join(projectRoot, mod);
      components.push(...this._scanProject(modRoot, visited));
    }

    return components;
  }

  // ─────────────────────────────
  // Merge duplicates by PURL
  // ─────────────────────────────
  static mergeInventoryByPurl(components) {
    const map = new Map();

    for (const comp of components) {
      if (!map.has(comp.id)) {
        map.set(comp.id, { ...comp, paths: [...comp.paths] });
        continue;
      }
      const existing = map.get(comp.id);
      for (const p of comp.paths)  { if (!existing.paths.includes(p))  existing.paths.push(p); }
      for (const s of comp.scopes) { if (!existing.scopes.includes(s)) existing.scopes.push(s); }
    }

    return [...map.values()];
  }

  // ─────────────────────────────
  // ENTRY
  // Walks startDir (and subdirs) for Maven project roots, then scans each.
  // A single shared `visited` set prevents any directory from being processed
  // more than once, regardless of whether it was found by the walk or by
  // multi-module recursion inside _scanProject.
  // ─────────────────────────────
  static async getInstalled(startDir = process.cwd()) {
    this.inventoryData = [];

    const visited = new Set();
    const raw     = [];

    // Scan startDir itself if it is a Maven root
    if (JavaMavenScanner._isMavenRoot(startDir)) {
      raw.push(...JavaMavenScanner._scanProject(startDir, visited));
    }

    // Walk subdirectories for additional Maven roots
    function walk(dir) {
      let entries;
      try {
        entries = fs.readdirSync(dir, { withFileTypes: true });
      } catch {
        return;
      }

      for (const entry of entries) {
        if (!entry.isDirectory()) continue;
        // Skip dirs that cannot contain user project files
        if (["node_modules", ".git", ".ubel", "target", ".m2"].includes(entry.name)) continue;

        const full = path.join(dir, entry.name);

        if (JavaMavenScanner._isMavenRoot(full)) {
          // _scanProject will add full to visited on first call;
          // multi-module recursion reuses the same set so no double-scan.
          raw.push(...JavaMavenScanner._scanProject(full, visited));
          // Do NOT recurse further into a Maven root — _scanProject follows
          // <modules> itself.
          continue;
        }

        walk(full);
      }
    }

    walk(startDir);

    const merged = this.mergeInventoryByPurl(raw);
    this.inventoryData = merged;
    return merged.map(c => c.id);
  }
}

export default JavaMavenScanner;

File: python/ubel/java_runner.py
"""
java_scanner.py — Java / Maven scanner (getInstalled only).

Mirrors the JS JavaMavenScanner:
  - Detects Maven project roots (pom.xml)
  - Reads .ubel/maven-deps.txt (mvn dependency:tree output) if present,
    otherwise parses pom.xml <dependency> blocks
  - Resolves ${property} version references from <properties>
  - Follows <modules> for multi-module builds (shared visited set)
  - Maps Maven scope → UBEL scope  (test → dev, rest → prod)
  - Merges duplicate PURLs
  - PURL: pkg:maven/<groupId>/<artifactId>@<version>
"""

from __future__ import annotations

import os
import re
from pathlib import Path
from typing import Any, Dict, List, Optional, Set


class JavaMavenScanner:

    # ------------------------------------------------------------------ #
    # PURL                                                                 #
    # ------------------------------------------------------------------ #

    
    def _maven_purl(self,group_id: str, artifact_id: str, version: str) -> str:
        base = f"pkg:maven/{group_id.lower()}/{artifact_id.lower()}"
        return f"{base}@{version}" if version else f"{base}@"

    # ------------------------------------------------------------------ #
    # Detect Maven root                                                    #
    # ------------------------------------------------------------------ #

    
    def _is_maven_root(self,directory: str) -> bool:
        return os.path.exists(os.path.join(directory, "pom.xml"))

    # ------------------------------------------------------------------ #
    # Minimal XML tag extractor                                            #
    # ------------------------------------------------------------------ #

    
    def _extract_tag(self,xml: str, tag: str) -> List[str]:
        return re.findall(
            rf"<{re.escape(tag)}[^>]*>([^<]*)</{re.escape(tag)}>",
            xml,
            re.IGNORECASE,
        )

    # ------------------------------------------------------------------ #
    # Parse <dependency> blocks from pom.xml                              #
    # ------------------------------------------------------------------ #

    
    def _parse_pom_deps(self,xml: str) -> List[Dict[str, str]]:
        deps: List[Dict[str, str]] = []
        for block_m in re.finditer(r"<dependency>([\s\S]*?)</dependency>", xml, re.IGNORECASE):
            block = block_m.group(1)
            group_id    = (self._extract_tag(block, "groupId")    or [""])[0]
            artifact_id = (self._extract_tag(block, "artifactId") or [""])[0]
            version     = (self._extract_tag(block, "version")    or [""])[0]
            scope       = (self._extract_tag(block, "scope")      or ["compile"])[0]
            optional    = (self._extract_tag(block, "optional")   or ["false"])[0]
            if group_id and artifact_id and optional.strip().lower() != "true":
                deps.append({
                    "groupId":    group_id.strip(),
                    "artifactId": artifact_id.strip(),
                    "version":    version.strip(),
                    "scope":      scope.strip().lower(),
                })
        return deps

    # ------------------------------------------------------------------ #
    # Parse <properties>                                                   #
    # ------------------------------------------------------------------ #

    
    def _parse_pom_properties(self,xml: str) -> Dict[str, str]:
        props: Dict[str, str] = {}
        props_m = re.search(r"<properties>([\s\S]*?)</properties>", xml, re.IGNORECASE)
        if not props_m:
            return props
        block = props_m.group(1)
        for m in re.finditer(r"<([A-Za-z0-9._-]+)>([^<]*)</[A-Za-z0-9._-]+>", block):
            props[m.group(1)] = m.group(2).strip()
        return props

    # ------------------------------------------------------------------ #
    # Resolve ${...} property refs                                         #
    # ------------------------------------------------------------------ #

    
    def _resolve_version(self,version: str, props: Dict[str, str]) -> str:
        if not version:
            return ""
        resolved = re.sub(
            r"\$\{([^}]+)\}",
            lambda m: props.get(m.group(1), ""),
            version,
        )
        return "" if "${" in resolved else resolved

    # ------------------------------------------------------------------ #
    # Parse mvn dependency:tree output                                     #
    # ------------------------------------------------------------------ #

    
    def _parse_deps_tree(self,file_path: str) -> List[Dict[str, str]]:
        deps: List[Dict[str, str]] = []
        try:
            content = Path(file_path).read_text(encoding="utf-8")
        except OSError:
            return deps

        for raw_line in content.splitlines():
            line = re.sub(r"^\[INFO\]\s*", "", raw_line)
            line = re.sub(r"^[|\s\\+\-]+", "", line).strip()
            # groupId:artifactId:packaging:version:scope
            m = re.match(r"^([^:]+):([^:]+):[^:]+:([^:]+):([^:\s]+)", line)
            if m:
                deps.append({
                    "groupId":    m.group(1),
                    "artifactId": m.group(2),
                    "version":    m.group(3),
                    "scope":      m.group(4).lower(),
                })
        return deps

    # ------------------------------------------------------------------ #
    # Parse <modules>                                                      #
    # ------------------------------------------------------------------ #

    
    def _parse_modules(self,xml: str) -> List[str]:
        mods: List[str] = []
        mods_m = re.search(r"<modules>([\s\S]*?)</modules>", xml, re.IGNORECASE)
        if not mods_m:
            return mods
        for m in re.finditer(r"<module>([^<]+)</module>", mods_m.group(1), re.IGNORECASE):
            mods.append(m.group(1).strip())
        return mods

    # ------------------------------------------------------------------ #
    # Maven scope → UBEL scope                                             #
    # ------------------------------------------------------------------ #

    
    def _maven_scope_to_ubel(self,mvn_scope: str) -> str:
        return "dev" if mvn_scope == "test" else "prod"

    # ------------------------------------------------------------------ #
    # Scan one Maven project                                               #
    # ------------------------------------------------------------------ #

    def _scan_project(self, project_root: str, visited: Set[str]) -> List[Dict[str, Any]]:
        key = os.path.realpath(project_root)
        if key in visited:
            return []
        visited.add(key)

        pom_path = os.path.join(project_root, "pom.xml")
        try:
            xml = Path(pom_path).read_text(encoding="utf-8")
        except OSError:
            return []

        # Strip XML comments
        xml = re.sub(r"<!--[\s\S]*?-->", "", xml)

        props      = self._parse_pom_properties(xml)
        components: List[Dict[str, Any]] = []

        # Prefer pre-generated dependency tree
        tree_file = os.path.join(project_root, ".ubel", "maven-deps.txt")
        tree_deps = self._parse_deps_tree(tree_file)

        if tree_deps:
            deps = tree_deps
        else:
            deps = [
                {**d, "version": self._resolve_version(d["version"], props)}
                for d in self._parse_pom_deps(xml)
            ]

        for dep in deps:
            if not dep.get("groupId") or not dep.get("artifactId"):
                continue
            version = dep.get("version") or ""
            if version == "unknown":
                version = ""
            cid   = self._maven_purl(dep["groupId"], dep["artifactId"], version)
            scope = self._maven_scope_to_ubel(dep.get("scope") or "compile")

            components.append({
                "id":           cid,
                "name":         f"{dep['groupId'].lower()}:{dep['artifactId'].lower()}",
                "version":      version,
                "type":         "library",
                "license":      "unknown",
                "ecosystem":    "java",
                "state":        "undetermined",
                "scopes":       [scope],
                "dependencies": [],
                "paths":        [project_root],
                "project_root": project_root,
            })

        # Recurse into <modules>
        for mod in self._parse_modules(xml):
            mod_root = os.path.join(project_root, mod)
            components.extend(self._scan_project(mod_root, visited))

        return components

    # ------------------------------------------------------------------ #
    # Merge by PURL                                                        #
    # ------------------------------------------------------------------ #

    
    def merge_inventory_by_purl(self,components: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        merged: Dict[str, Dict] = {}
        for comp in components:
            cid = comp["id"]
            if cid not in merged:
                clone = dict(comp)
                clone["paths"]  = list(comp.get("paths", []))
                clone["scopes"] = list(comp.get("scopes", []))
                merged[cid] = clone
                continue
            existing = merged[cid]
            for p in comp.get("paths", []):
                if p not in existing["paths"]:
                    existing["paths"].append(p)
            for s in comp.get("scopes", []):
                if s not in existing["scopes"]:
                    existing["scopes"].append(s)
        return list(merged.values())

    # ------------------------------------------------------------------ #
    # ENTRY                                                                #
    # ------------------------------------------------------------------ #

    def get_installed(self, start_dir: str = ".") -> List[str]:
        self.inventory_data = []
        start_dir = os.path.abspath(start_dir)

        visited: Set[str] = set()
        raw: List[Dict[str, Any]] = []

        if self._is_maven_root(start_dir):
            raw.extend(self._scan_project(start_dir, visited))

        skip = {"node_modules", ".git", ".ubel", "target", ".m2"}

        def walk(directory: str) -> None:
            try:
                entries = os.scandir(directory)
            except OSError:
                return
            with entries as it:
                for entry in it:
                    if not entry.is_dir(follow_symlinks=False):
                        continue
                    if entry.name in skip:
                        continue
                    full = entry.path
                    if self._is_maven_root(full):
                        raw.extend(self._scan_project(full, visited))
                        # Don't descend further — _scan_project follows <modules>
                        continue
                    walk(full)

        walk(start_dir)

        merged = self.merge_inventory_by_purl(raw)
        self.inventory_data = merged
        return [c["id"] for c in merged]